File: exercise/98/solution2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        stack<pair<TreeNode*, pair<long, long>>> nodes;
        nodes.push(make_pair(root, make_pair(LONG_MIN, LONG_MAX)));
        while (!nodes.empty()) {
            auto now = nodes.top();
            nodes.pop();
            TreeNode *p = now.first;
            if (!p) continue;
            auto limit = now.second;
            if (p->val <= limit.first || p->val >= limit.second) return false;
            nodes.push(make_pair(p->left, make_pair(limit.first, p->val)));
            nodes.push(make_pair(p->right, make_pair(p->val, limit.second)));
        }
        return true;
    }
};

File: exercise/98/solution2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def isValidBSTEdge(self, root: TreeNode, left: int, right: int) -> bool:
        if root is None:
            return True
        if left is not None and left >= root.val:
            return False
        if right is not None and right <= root.val:
            return False
        left_valid = self.isValidBSTEdge(root.left, left, root.val)
        right_valid = self.isValidBSTEdge(root.right, root.val, right)
        return left_valid and right_valid

    def isValidBST(self, root: TreeNode) -> bool:
        return self.isValidBSTEdge(root, None, None)
